src/execution.c

```diff
@@ -92,21 +104,39 @@ int	execution(t_macro *macro)
 	int		read_end;
 	int		num_cmds_executed;
 	char	**cmd_array;
+	int	 	pipe_exit[2];
+	int 	i;
+	int 	status;
 
 	if (macro->num_cmds == 1 && macro->cmds->type == BUILTIN)
 	{
 		cmd_array = build_cmd_args_array(macro->cmds->cmd_arg);
 		if (cmd_array == NULL)
-			return (-1);
+			return (1);
 		g_exit = execute_builtin(macro, cmd_array); // wait and pid ???
 		free_array(&cmd_array);
 	}
 	else
 	{
+		if (pipe(pipe_exit) == -1)
+			return (error_msg("pipe failed", 0));
 		read_end = 0;
 		macro->pid = malloc(sizeof(pid_t) * macro->num_cmds);
-		num_cmds_executed = execute_cmds(macro, read_end);
-		g_exit = wait_processes(macro->pid, num_cmds_executed);
+		num_cmds_executed = execute_cmds(macro, read_end, pipe_exit);
+		i = 0;
+		while (i < num_cmds_executed)
```

Comment:
Why not use the wait process and get_exit functions? you just used them inside this function. I have not tested, but as you commented that you are still testing for catching the **last** exit code, seems natural to use those functions. They have been tested in pipex already. I suggest adding the pipe closing step inside the waiting function **after** the last has been found. I'm not sure if the get_exit status help us catching the error codes with the signals, but maybe also is a good idea.

Reply:
I was first using those two functions and had some problems (didn't return what it was supposed to), so I just replaced them. I need to look into it because I will need to use external functions anyways due to the 25-line per function limit.